
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../examples/scdef-ifn/">
      
      
      
      <link rel="icon" href="../assets/images/icon.svg">
      <meta name="generator" content="mkdocs-1.6.0, mkdocs-material-9.5.30">
    
    
      
        <title>API Reference - scDEF</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.3cba04c6.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#api-reference" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="scDEF" class="md-header__button md-logo" aria-label="scDEF" data-md-component="logo">
      
  <img src="../assets/images/icon.svg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            scDEF
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              API Reference
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/cbg-ethz/scDEF" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    scDEF
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href=".." class="md-tabs__link">
        
  
    
  
  Getting started

      </a>
    </li>
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../examples/basicusage/" class="md-tabs__link">
          
  
  Examples

        </a>
      </li>
    
  

      
        
  
  
    
  
  
    <li class="md-tabs__item md-tabs__item--active">
      <a href="./" class="md-tabs__link">
        
  
    
  
  API Reference

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


  

<nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="scDEF" class="md-nav__button md-logo" aria-label="scDEF" data-md-component="logo">
      
  <img src="../assets/images/icon.svg" alt="logo">

    </a>
    scDEF
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/cbg-ethz/scDEF" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    scDEF
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Getting started
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Examples
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Examples
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../examples/basicusage/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Basic usage
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../examples/scdef-pbmcs3k/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    scDEF on 3k PBMCs
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../examples/scdef-planaria/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Cell type hierarchies in a whole adult animal
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../examples/scdef-pbmcs-2batches/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    scDEF on two batches of PBMCs
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../examples/scdef-ifn/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Identifying interferon-response between two batches of PBMCs
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    API Reference
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    API Reference
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#model-classes" class="md-nav__link">
    <span class="md-ellipsis">
      Model classes
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#scdef.scDEF" class="md-nav__link">
    <span class="md-ellipsis">
      scDEF
    </span>
  </a>
  
    <nav class="md-nav" aria-label="scDEF">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#scdef.scDEF.filter_factors" class="md-nav__link">
    <span class="md-ellipsis">
      filter_factors
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#scdef.scDEF.get_hierarchy" class="md-nav__link">
    <span class="md-ellipsis">
      get_hierarchy
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#scdef.scDEF.learn" class="md-nav__link">
    <span class="md-ellipsis">
      learn
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#scdef.scDEF.make_graph" class="md-nav__link">
    <span class="md-ellipsis">
      make_graph
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#scdef.scDEF.plot_multilevel_paga" class="md-nav__link">
    <span class="md-ellipsis">
      plot_multilevel_paga
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#scdef.scDEF.plot_obs_scores" class="md-nav__link">
    <span class="md-ellipsis">
      plot_obs_scores
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#scdef.scDEF.plot_pathway_scores" class="md-nav__link">
    <span class="md-ellipsis">
      plot_pathway_scores
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#scdef.scDEF.plot_signatures_scores" class="md-nav__link">
    <span class="md-ellipsis">
      plot_signatures_scores
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#scdef.iscDEF" class="md-nav__link">
    <span class="md-ellipsis">
      iscDEF
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#benchmarking-utilities" class="md-nav__link">
    <span class="md-ellipsis">
      Benchmarking utilities
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#scdef.benchmark.run_multiple_resolutions" class="md-nav__link">
    <span class="md-ellipsis">
      run_multiple_resolutions
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                



                  

  
  


<h1 id="api-reference">API Reference</h1>
<h2 id="model-classes">Model classes</h2>


<div class="doc doc-object doc-class">



<h2 id="scdef.scDEF" class="doc doc-heading">
          <code>scdef.scDEF</code>


</h2>


  <div class="doc doc-contents first">
          <p class="doc doc-class-bases">
            Bases: <code>object</code></p>

  
      <p>Single-cell Deep Exponential Families model.</p>
<p>This model learns multi-level gene signatures describing the input scRNA-seq
data from an AnnData object.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>adata</code></td>
          <td>
                <code><span title="anndata.AnnData">AnnData</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>AnnData object containing the gene expression data. scDEF learns a model from
counts, so they must be present in either adata.X or in adata.layers.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>counts_layer</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[str]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>layer from adata.layers to get the count data from.</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
        <tr>
          <td><code>layer_sizes</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[list]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>number of factors per scDEF layer.</p>
            </div>
          </td>
          <td>
                <code>[100, 60, 30, 10, 1]</code>
          </td>
        </tr>
        <tr>
          <td><code>batch_key</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[str]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>key in adata.obs containing batch annotations for batch correction. If None, or not found,
no batch correction is performed.</p>
            </div>
          </td>
          <td>
                <code>&#39;batch&#39;</code>
          </td>
        </tr>
        <tr>
          <td><code>seed</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[int]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>random seed for JAX</p>
            </div>
          </td>
          <td>
                <code>1</code>
          </td>
        </tr>
        <tr>
          <td><code>logginglevel</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[int]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>verbosity level for logger</p>
            </div>
          </td>
          <td>
                <code><span title="logging.INFO">INFO</span></code>
          </td>
        </tr>
        <tr>
          <td><code>layer_shapes</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[list]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>prior parameters for the z shape to use in each scDEF layer</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
        <tr>
          <td><code>brd_strength</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[float]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>BRD prior concentration parameter</p>
            </div>
          </td>
          <td>
                <code>1000.0</code>
          </td>
        </tr>
        <tr>
          <td><code>brd_mean</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[float]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>BRD prior mean parameter</p>
            </div>
          </td>
          <td>
                <code>0.01</code>
          </td>
        </tr>
        <tr>
          <td><code>use_brd</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[bool]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>whether to use the BRD prior for factor relevance estimation</p>
            </div>
          </td>
          <td>
                <code>True</code>
          </td>
        </tr>
        <tr>
          <td><code>cell_scale_shape</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[float]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>concentration level in the cell scale prior</p>
            </div>
          </td>
          <td>
                <code>1.0</code>
          </td>
        </tr>
        <tr>
          <td><code>gene_scale_shape</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[float]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>concentration level in the gene scale prior</p>
            </div>
          </td>
          <td>
                <code>1.0</code>
          </td>
        </tr>
        <tr>
          <td><code>factor_shapes</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[list]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>prior parameters for the W shape to use in each scDEF layer</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
        <tr>
          <td><code>factor_rates</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[list]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>prior parameters for the W rate to use in each scDEF layer</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
        <tr>
          <td><code>layer_diagonals</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[list]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>prior diagonal strengths for the W parameters in each scDEF layer</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
        <tr>
          <td><code>batch_cpal</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[str]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>default color palette for batch annotations</p>
            </div>
          </td>
          <td>
                <code>&#39;Dark2&#39;</code>
          </td>
        </tr>
        <tr>
          <td><code>layer_cpal</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[list]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>default color palettes for scDEF layers</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
        <tr>
          <td><code>lightness_mult</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[float]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>multiplier to define lightness of color palette at each scDEF layer</p>
            </div>
          </td>
          <td>
                <code>0.15</code>
          </td>
        </tr>
    </tbody>
  </table>


  

  <div class="doc doc-children">










<div class="doc doc-object doc-function">



<h3 id="scdef.scDEF.filter_factors" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">filter_factors</span><span class="p">(</span><span class="n">thres</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">iqr_mult</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mf">0.005</span><span class="p">,</span> <span class="n">filter_up</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Filter our irrelevant factors based on the BRD posterior or the cell attachments.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>thres</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[float]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>minimum factor BRD value</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
        <tr>
          <td><code>iqr_mult</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[float]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>multiplier of the difference between the third quartile and the median BRD values to set the threshold</p>
            </div>
          </td>
          <td>
                <code>0.0</code>
          </td>
        </tr>
        <tr>
          <td><code>min_cells</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[float]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>minimum number of cells that each factor must have attached to it for it to be kept. If between 0 and 1, fraction. Otherwise, absolute value</p>
            </div>
          </td>
          <td>
                <code>0.005</code>
          </td>
        </tr>
        <tr>
          <td><code>filter_up</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[bool]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>whether to remove factors in upper layers via inter-layer attachments</p>
            </div>
          </td>
          <td>
                <code>True</code>
          </td>
        </tr>
    </tbody>
  </table>

  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="scdef.scDEF.get_hierarchy" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_hierarchy</span><span class="p">(</span><span class="n">simplified</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Get a dictionary containing the polytree contained in the scDEF graph.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>simplified</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[bool]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>whether to collapse single-child nodes</p>
            </div>
          </td>
          <td>
                <code>True</code>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
<th>Name</th>        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
<td><code>hierarchy</code></td>          <td>
                <code><span title="typing.Mapping">Mapping</span>[str, <span title="typing.Sequence">Sequence</span>[str]]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>the dictionary containing the hierarchy</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="scdef.scDEF.learn" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">learn</span><span class="p">(</span><span class="n">n_epoch</span><span class="o">=</span><span class="p">[</span><span class="mi">1000</span><span class="p">,</span> <span class="mi">1000</span><span class="p">],</span> <span class="n">lr</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">annealing</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">num_samples</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">batch_size</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">layerwise</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Fit a variational approximation to the posterior over scDEF parameters.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>n_epoch</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<span title="typing.Union">Union</span>[int, list]]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>number of epochs (full passes of the data).
Can be a list of ints for multi-step learning.</p>
            </div>
          </td>
          <td>
                <code>[1000, 1000]</code>
          </td>
        </tr>
        <tr>
          <td><code>lr</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<span title="typing.Union">Union</span>[float, list]]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>learning rate.
Can be a list of floats for multi-step learning.</p>
            </div>
          </td>
          <td>
                <code>0.1</code>
          </td>
        </tr>
        <tr>
          <td><code>annealing</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<span title="typing.Union">Union</span>[float, list]]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>scale factor for the entropy term.
Can be a list of floats for multi-step learning.</p>
            </div>
          </td>
          <td>
                <code>1.0</code>
          </td>
        </tr>
        <tr>
          <td><code>num_samples</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[int]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>number of Monte Carlo samples to use in the ELBO approximation.</p>
            </div>
          </td>
          <td>
                <code>10</code>
          </td>
        </tr>
        <tr>
          <td><code>batch_size</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[int]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>number of data points to use per iteration. If None, uses all.
Useful for data sets that do not fit in GPU memory.</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
        <tr>
          <td><code>layerwise</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[bool]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>whether to optimize the model parameters in a step-wise manner:
first learn only Layer 0 and 1, and then 2, and then 3, and so on. The size of
the n_epoch or lr schedules will be ignored, only the first value will be used
and each step will use that n_epoch value.</p>
            </div>
          </td>
          <td>
                <code>False</code>
          </td>
        </tr>
    </tbody>
  </table>

  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="scdef.scDEF.make_graph" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">make_graph</span><span class="p">(</span><span class="n">hierarchy</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">show_all</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">factor_annotations</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">top_factor</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">show_signatures</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">enrichments</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">top_genes</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">show_batch_counts</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">filled</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">wedged</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">color_edges</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">show_confidences</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">mc_samples</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">n_cells_label</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">n_cells</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">node_size_max</span><span class="o">=</span><span class="mf">2.0</span><span class="p">,</span> <span class="n">node_size_min</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">scale_level</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">show_label</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">gene_score</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">gene_cmap</span><span class="o">=</span><span class="s1">&#39;viridis&#39;</span><span class="p">,</span> <span class="o">**</span><span class="n">fontsize_kwargs</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Make Graphviz-formatted scDEF graph.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>hierarchy</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[dict]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>a dictionary containing the polytree to draw instead of the whole graph</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
        <tr>
          <td><code>show_all</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[bool]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>whether to show all factors even post filtering</p>
            </div>
          </td>
          <td>
                <code>False</code>
          </td>
        </tr>
        <tr>
          <td><code>factor_annotations</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[dict]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>factor annotations to include in the node labels</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
        <tr>
          <td><code>top_factor</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[str]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>only include factors below this factor</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
        <tr>
          <td><code>show_signatures</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[bool]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>whether to show the ranked gene signatures in the node labels</p>
            </div>
          </td>
          <td>
                <code>True</code>
          </td>
        </tr>
        <tr>
          <td><code>enrichments</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<span title="pandas.DataFrame">DataFrame</span>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>enrichment results from gseapy to include in the node labels</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
        <tr>
          <td><code>top_genes</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[int]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>number of genes from each signature to be shown in the node labels</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
        <tr>
          <td><code>show_batch_counts</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[bool]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>whether to show the number of cells from each batch that attach to each factor</p>
            </div>
          </td>
          <td>
                <code>False</code>
          </td>
        </tr>
        <tr>
          <td><code>filled</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[str]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>key from self.adata.obs to use to fill the nodes with</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
        <tr>
          <td><code>wedged</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[str]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>key from self.adata.obs to use to wedge the nodes with</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
        <tr>
          <td><code>color_edges</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[bool]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>whether to color the graph edges according to the upper factors</p>
            </div>
          </td>
          <td>
                <code>True</code>
          </td>
        </tr>
        <tr>
          <td><code>show_confidences</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[bool]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>whether to show the confidence score for each signature</p>
            </div>
          </td>
          <td>
                <code>False</code>
          </td>
        </tr>
        <tr>
          <td><code>mc_samples</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[int]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>number of Monte Carlo samples to take from the posterior to compute signature confidences</p>
            </div>
          </td>
          <td>
                <code>100</code>
          </td>
        </tr>
        <tr>
          <td><code>n_cells_label</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[bool]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>wether to show the number of cells that attach to the factor</p>
            </div>
          </td>
          <td>
                <code>False</code>
          </td>
        </tr>
        <tr>
          <td><code>n_cells</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[bool]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>wether to scale the node sizes by the number of cells that attach to the factor</p>
            </div>
          </td>
          <td>
                <code>False</code>
          </td>
        </tr>
        <tr>
          <td><code>node_size_max</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[int]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>maximum node size when scaled by cell numbers</p>
            </div>
          </td>
          <td>
                <code>2.0</code>
          </td>
        </tr>
        <tr>
          <td><code>node_size_min</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[int]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>minimum node size when scaled by cell numbers</p>
            </div>
          </td>
          <td>
                <code>0.05</code>
          </td>
        </tr>
        <tr>
          <td><code>scale_level</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[bool]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>wether to scale node sizes per level instead of across all levels</p>
            </div>
          </td>
          <td>
                <code>False</code>
          </td>
        </tr>
        <tr>
          <td><code>show_label</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[bool]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>wether to show labels on nodes</p>
            </div>
          </td>
          <td>
                <code>True</code>
          </td>
        </tr>
        <tr>
          <td><code>gene_score</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[str]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>color the nodes by the score they attribute to a gene, normalized by layer. Overrides filled and wedged</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
        <tr>
          <td><code>gene_cmap</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[str]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>colormap to use for gene_score</p>
            </div>
          </td>
          <td>
                <code>&#39;viridis&#39;</code>
          </td>
        </tr>
        <tr>
          <td><code>**fontsize_kwargs</code></td>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>keyword arguments to adjust the fontsizes according to the gene scores</p>
            </div>
          </td>
          <td>
                <code>{}</code>
          </td>
        </tr>
    </tbody>
  </table>

  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="scdef.scDEF.plot_multilevel_paga" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">plot_multilevel_paga</span><span class="p">(</span><span class="n">neighbors_rep</span><span class="o">=</span><span class="s1">&#39;X_factors&#39;</span><span class="p">,</span> <span class="n">layers</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">16</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span> <span class="n">reuse_pos</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">paga_kwargs</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Plot a PAGA graph from each scDEF layer.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>neighbors_rep</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[str]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>the self.obsm key to use to compute the PAGA graphs</p>
            </div>
          </td>
          <td>
                <code>&#39;X_factors&#39;</code>
          </td>
        </tr>
        <tr>
          <td><code>layers</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[list]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>which layers to plot</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
        <tr>
          <td><code>figsize</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[tuple]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>figure size</p>
            </div>
          </td>
          <td>
                <code>(16, 4)</code>
          </td>
        </tr>
        <tr>
          <td><code>reuse_pos</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[bool]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>whether to initialize each PAGA graph with the graph from the layer above</p>
            </div>
          </td>
          <td>
                <code>True</code>
          </td>
        </tr>
        <tr>
          <td><code>show</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[bool]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>whether to show the plot</p>
            </div>
          </td>
          <td>
                <code>True</code>
          </td>
        </tr>
        <tr>
          <td><code>**paga_kwargs</code></td>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>keyword arguments to adjust the PAGA layouts</p>
            </div>
          </td>
          <td>
                <code>{}</code>
          </td>
        </tr>
    </tbody>
  </table>

  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="scdef.scDEF.plot_obs_scores" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">plot_obs_scores</span><span class="p">(</span><span class="n">obs_keys</span><span class="p">,</span> <span class="n">hierarchy</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;fracs&#39;</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Plot the association between a set of cell annotations and factors.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>obs_keys</code></td>
          <td>
                <code><span title="typing.Sequence">Sequence</span>[str]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>the keys in self.adata.obs to use</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>hierarchy</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[dict]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>the polytree to restrict the associations to</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
        <tr>
          <td><code>mode</code></td>
          <td>
                <code><span title="typing.Literal">Literal</span>[&#39;f1&#39;, &#39;fracs&#39;, &#39;weights&#39;]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>whether to compute scores based on assignments or weights</p>
            </div>
          </td>
          <td>
                <code>&#39;fracs&#39;</code>
          </td>
        </tr>
        <tr>
          <td><code>**kwargs</code></td>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>plotting keyword arguments</p>
            </div>
          </td>
          <td>
                <code>{}</code>
          </td>
        </tr>
    </tbody>
  </table>

  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="scdef.scDEF.plot_pathway_scores" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">plot_pathway_scores</span><span class="p">(</span><span class="n">pathways</span><span class="p">,</span> <span class="n">top_genes</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Plot the association between a set of cell annotations and a set of gene signatures.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>obs_keys</code></td>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>the keys in self.adata.obs to use</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>pathways</code></td>
          <td>
                <code><span title="pandas.DataFrame">DataFrame</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>a pandas DataFrame containing PROGENy pathways</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>**kwargs</code></td>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>plotting keyword arguments</p>
            </div>
          </td>
          <td>
                <code>{}</code>
          </td>
        </tr>
    </tbody>
  </table>

  </div>

</div>


<div class="doc doc-object doc-function">



<h3 id="scdef.scDEF.plot_signatures_scores" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">plot_signatures_scores</span><span class="p">(</span><span class="n">obs_keys</span><span class="p">,</span> <span class="n">markers</span><span class="p">,</span> <span class="n">top_genes</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">hierarchy</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Plot the association between a set of cell annotations and a set of gene signatures.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>obs_keys</code></td>
          <td>
                <code><span title="typing.Sequence">Sequence</span>[str]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>the keys in self.adata.obs to use</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>markers</code></td>
          <td>
                <code><span title="typing.Mapping">Mapping</span>[str, <span title="typing.Sequence">Sequence</span>[str]]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>a dictionary with keys corresponding to self.adata.obs[obs_keys] and values to gene lists</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>top_genes</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[int]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>number of genes to consider in the score computations</p>
            </div>
          </td>
          <td>
                <code>10</code>
          </td>
        </tr>
        <tr>
          <td><code>hierarchy</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[dict]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>the polytree to restrict the associations to</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
        <tr>
          <td><code>**kwargs</code></td>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>plotting keyword arguments</p>
            </div>
          </td>
          <td>
                <code>{}</code>
          </td>
        </tr>
    </tbody>
  </table>

  </div>

</div>



  </div>

  </div>


</div>

<div class="doc doc-object doc-class">



<h2 id="scdef.iscDEF" class="doc doc-heading">
          <code>scdef.iscDEF</code>


</h2>


  <div class="doc doc-contents first">
          <p class="doc doc-class-bases">
            Bases: <code><a class="autorefs autorefs-internal" title="scdef.models._scdef.scDEF" href="#scdef.scDEF">scDEF</a></code></p>

  
      <p>Informed scDEF model.</p>
<p>This model extends the basic scDEF by using gene sets to guide the factors.
iscDEF can either set the given sets as top layer factors and learn higher-resolution
structure, or use them as the lowest resolution and learn a hierarchy that relates them.
All the methods from scDEF are available in iscDEF.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>adata</code></td>
          <td>
                <code><span title="anndata.AnnData">AnnData</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>AnnData object containing the gene expression data. scDEF learns a model from
counts, so they must be present in either adata.X or in adata.layers.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>markers_dict</code></td>
          <td>
                <code><span title="typing.Mapping">Mapping</span>[str, <span title="typing.Sequence">Sequence</span>[str]]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>dictionary containing named gene lists.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>add_other</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[bool]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>whether to add factors for cells which don't express any of the sets in markers_dict.</p>
            </div>
          </td>
          <td>
                <code>False</code>
          </td>
        </tr>
        <tr>
          <td><code>markers_layer</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[int]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>scDEF layer at which the gene sets are defined. If &gt; 0, this defines the number of layers.</p>
            </div>
          </td>
          <td>
                <code>0</code>
          </td>
        </tr>
        <tr>
          <td><code>n_factors_per_set</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[int]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>number of lower level factors per gene set.</p>
            </div>
          </td>
          <td>
                <code>3</code>
          </td>
        </tr>
        <tr>
          <td><code>n_layers</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[int]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>default number of scDEF layers, including a top layer of size 1 if markers_layer is 0.</p>
            </div>
          </td>
          <td>
                <code>4</code>
          </td>
        </tr>
        <tr>
          <td><code>cn_small_scale</code></td>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>scale for low connectivity</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>cn_big_scale</code></td>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>scale for large connectivity</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>cn_small_strength</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[float]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>strength for weak connectivity</p>
            </div>
          </td>
          <td>
                <code>100.0</code>
          </td>
        </tr>
        <tr>
          <td><code>cn_big_strength</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[float]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>strength for large connectivity</p>
            </div>
          </td>
          <td>
                <code>1.0</code>
          </td>
        </tr>
        <tr>
          <td><code>gs_small_scale</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[float]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>scale for genes not in set</p>
            </div>
          </td>
          <td>
                <code>0.1</code>
          </td>
        </tr>
        <tr>
          <td><code>gs_big_scale</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[float]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>scale for genes in set</p>
            </div>
          </td>
          <td>
                <code>10.0</code>
          </td>
        </tr>
        <tr>
          <td><code>marker_strength</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[float]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>strength for marker genes</p>
            </div>
          </td>
          <td>
                <code>100.0</code>
          </td>
        </tr>
        <tr>
          <td><code>nonmarker_strength</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[float]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>strength for non-marker genes</p>
            </div>
          </td>
          <td>
                <code>1.0</code>
          </td>
        </tr>
        <tr>
          <td><code>other_strength</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[float]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>strength for marker genes of other sets</p>
            </div>
          </td>
          <td>
                <code>100.0</code>
          </td>
        </tr>
        <tr>
          <td><code>**kwargs</code></td>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>keyword arguments for base scDEF.</p>
            </div>
          </td>
          <td>
                <code>{}</code>
          </td>
        </tr>
    </tbody>
  </table>


  

  <div class="doc doc-children">











  </div>

  </div>


</div><h2 id="benchmarking-utilities">Benchmarking utilities</h2>



<div class="doc doc-object doc-function">



<h2 id="scdef.benchmark.run_multiple_resolutions" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">scdef</span><span class="o">.</span><span class="n">benchmark</span><span class="o">.</span><span class="n">run_multiple_resolutions</span><span class="p">(</span><span class="n">method</span><span class="p">,</span> <span class="n">ad</span><span class="p">,</span> <span class="n">resolution_sweep</span><span class="p">,</span> <span class="n">layer_prefix</span><span class="o">=</span><span class="s1">&#39;h&#39;</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents first">
  
      <p>Run a clustering and gene signature learning method at multiple resolutions.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>method</code></td>
          <td>
                <code><span title="typing.Callable">Callable</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>the function that runs the method. Must take a resolution parameter as argument and
return a list containing at least an AnnData object, a matrix containing the latent space,
and a list of genes per cluster.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>ad</code></td>
          <td>
                <code><span title="anndata.AnnData">AnnData</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>the data to run the method on.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>resolution_sweep</code></td>
          <td>
                <code><span title="typing.Sequence">Sequence</span>[float]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>list of resolution parameters to use.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
<th>Name</th>        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
<td><code>outs</code></td>          <td>
                <code><span title="typing.Mapping">Mapping</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>dictionary containing all the outputs from the method across all resolutions. Keys: ["latents", "signatures", "assignments", "scores", "sizes", "simplified_hierarchy"]</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

  </div>

</div>












                

              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var tab,labels=set.querySelector(".tabbed-labels");for(tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": ["navigation.tabs", "navigation.sections", "toc.integrate", "search.suggest", "search.highlight", "content.tabs.link", "content.code.annotation", "content.code.copy"], "search": "../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.fe8b6f2b.min.js"></script>
      
    
  </body>
</html>